### PR TITLE
Update GridSearch and add to tutorial

### LINKED ISSUE #436: GridSearch back up and running in a tutorial
Looks like most of the code is already there. Hopefully very little tweaking is required to get it in one of the tutorials again.

### PR DESCRIPTION
**Pull request by henryre**
Berg gets back on the board

Fixes #436

### CODE REVIEW

File: snorkel/learning.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def predict(self, X, b=0.5):
         """Return numpy array of elements in {-1,0,1} based on predicted marginal probabilities."""
         return np.array([1 if p > b else -1 if p < b else 0 for p in self.marginals(X)])
 
-    def score(self, X_test, L_test, gold_candidate_set, b=0.5, set_unlabeled_as_neg=True):
+    def score(self, X_test, L_test, gold_candidate_set, b=0.5, set_unlabeled_as_neg=True, disp=True):
```
Comment (ajratner) [inline snorkel/learning.py]: `display` for readability?

File: snorkel/learning_utils.py [hunk 2]
```diff
@@ -193,18 +193,23 @@ class GridSearch(object):
...
+        self.model              = model
+        self.X                  = X
+        self.training_marginals = training_marginals
+        self.param_names        = param_names
+        self.param_val_ranges   = param_val_ranges
+        
+    def search_space(self):
+        return product(*self.param_val_ranges)
+
+    def fit(self, X_test, L_test, gold_candidate_set, b=0.5, set_unlabeled_as_neg=True, **model_hyperparams):
```
Comment (ajratner) [inline snorkel/learning_utils.py]: Some minor things:
* I would call e.g. `X_validation` to make it clear that they shouldn't use the test set here?
* `L_test` is a vector of labels, so should be e.g. `validation_labels` (upper case single letter reserved for matrices)

File: snorkel/learning_utils.py [hunk 3]
```diff
@@ -214,24 +219,33 @@ def fit(self, candidates, gold_labels, **model_hyperparams):
...
             # Train the model
-            self.learner.train(**model_hyperparams)
+            self.model.train(self.X, self.training_marginals, **model_hyperparams)
 
             # Test the model
-            scores   = self.learner.test(candidates, gold_labels, display=False, return_vals=True)
-            p, r, f1 = scores[:3]
+            tp, fp, tn, fn = self.model.score(X_test, L_test, gold_candidate_set, b, set_unlabeled_as_neg, disp=False)
+            p, r = float(len(tp)) / (len(tp) + len(fp)), float(len(tp)) / (len(tp) + len(fn))
+            f1 = 2.0 * (p * r) / (p + r) if (p + r) > 0 else 0
```
Comment (ajratner) [inline snorkel/learning_utils.py]: We should put these as standard utility fns

File: snorkel/learning_utils.py [hunk 4]
```diff
@@ -214,24 +219,33 @@ def fit(self, candidates, gold_labels, **model_hyperparams):
...
-        self.learner.model.w = w_opt
+        self.model.w = w_opt
 
         # Return DataFrame of scores
         self.results = DataFrame.from_records(run_stats, columns=self.param_names + ['Prec.', 'Rec.', 'F1'])
         return self.results
-
+    
+    
+class RandomSearch(GridSearch):
```
Comment (ajratner) [inline snorkel/learning_utils.py]: Minor: I was thinking `RandomSearch` would be over a continuous space of param values, rather than filtering a set of given discreet values as here; I think that would be more convenient because then the user only needs to specify a range rather than a set of values?

Comment (henryre) [inline snorkel/learning_utils.py]: Shhhhhhhhh values

**Comment by henryre** [comment]
Ok @ajratner changes made

File: snorkel/learning_utils.py (hunk 4 above)
Comment (ajratner) [inline snorkel/learning_utils.py]: Well played

**Comment by ajratner** [comment]
@henryre looks good, added a bit of text in tutorial.  Two questions remaining:

- What's default if they don't have any dev labels?  What were you doing before (we can discuss this in person)
- Did you check to make sure adding this in preserves or improves the tutorial score (if not, we need to debug)?

**Comment by henryre** [comment]
@ajratner 

- Let's add the no labels option later. Needs a discussion about best practices. Previous solution was: pick a random hold out set, then train the models and select the w s.t. sqrt(np.mean((p - 0.5)^2)) is maximized where p were the predicted marginals for the hold out set (i.e. measures how spread out the marginals are)
- 90% certain it does improve performance (from memory), but will double check shortly

**Comment by henryre** [comment]
lol progressbar broke the build

**Comment by stephenbach** [comment]
Might be helpful. Same issue: https://github.com/fchollet/keras/issues/2110

**Comment by ajratner** [comment]
So let's just put in how to proceed without grid search (if you don't have dev set labeled)- this will preempt a lot of questions!

**Comment by henryre** [comment]
Dev set F1: 0.4 -> 0.57
Test set F1: 0.7 -> 0.63

**Comment by ajratner** [comment]
LGTM- will merge in after Travis passes; we should merge into `dev` and `master`
